src/mds/CDir.cc

```diff
@@ -2083,10 +2083,13 @@ void CDir::go_bad(bool complete)
     ceph_abort();  // unreachable, damaged() respawns us
   }
 
-  if (complete)
+  if (complete) {
     _go_bad();
```

Comment:
While you're here, I'm not sure why `_go_bad` is a helper method. How about move it into this `if` and call `state_clear(STATE_FETCHING)` before this `if` since both paths need to clear `STATE_FETCHING`. Both paths also need to do `finish_waiting(WAIT_COMPLETE, -EIO)`, which can go below the `else`.

Reply:
Yes, current fix has some duplicated codes. According to your suggestion, I got two ideas:

1. Since `_go_bad` is already a helper method, how about we move all the stuff including checking `complete` flag into `_go_bad`.

2.  Get rid of `_go_bad` method, just do all the stuff in `go_bad`, since `_go_bad` is only called by `go_bad` right now.

I think it's better not to move `state_clear(STATE_FETCHING)` and `finish_waiting(WAIT_COMPLETE, -EIO)` out of `_go_bad`, in case some one use `_go_bad` later but forget to do `state_clear` or `finish_waiting`.

Reply:
@batrick  Pls check my last comment. Which one do you prefer?

Reply:
(2)